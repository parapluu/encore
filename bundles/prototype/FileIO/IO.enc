bundle FileIO.IO where

passive class File {

  file   : embed FILE* end
  fname  : String
  mode   : String

  def init(fname:String, mode:String) : void {
    this.fname  = fname;
    this.mode   = mode;
  }

  def open(fin:String, mode:String) : void {
    this.mode = if (mode.equals("")) then "w+r+a"
                else mode;
    this.fname= fin;

    let fin   = fin.data;
    let mChar = this.mode.data;
    this.file = embed (embed FILE* end)
                  FILE *file = fopen(#{fin}, #{mChar});
                  if (!file) {
                    printf("Cannot open file, exiting.\n");
                    exit(1);
                  };
                  file;
                end;
  }

  def close() : void {
    let f = this.file;
    this.file = embed (embed FILE* end)
      if (!#{f}) {
        printf("Cannot close file, exiting.\n");
        exit(1);
      };
      fclose(#{f});
      (FILE*)NULL;
    end;
  }

  -----------------------------------
  -- Write to file.
  -----------------------------------
  def write(content:String) : void {
    this.writeChar(content.data);
  }

  def writeChar(content:embed char* end) : void {
    let file  = this.file;
    embed void
      FILE *fout = #{file};
      if (!fout) {
        printf("Cannot open file, exiting.\n");
        exit(1);
      }
      fprintf(fout, #{content});
    end
  }

  -----------------------------------
  -- Read file.
  -----------------------------------
  def readline() : String {
    new String(this.readlineChar());
  }

  def readlineChar() : embed char* end {
    let f = this.file;
    embed (embed char* end)
      char* line = encore_alloc(*_ctx,255);
      fgets(line, 255, #{f});
      line;
    end;
  }

  def eof() : bool {
    let f = this.file;
    embed bool feof(#{f}); end
  }
}