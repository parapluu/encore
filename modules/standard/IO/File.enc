module File

import Data.StringBuffer

typedef FilePath = String

typedef FILE = EMBED FILE* END


local class File
  var file : FILE              -- convert to Maybe to avoid using null pointer under the hood
  val file_name : String
  val mode : String            -- replace by IOMode ADT

  def init(fname : FilePath, mode : String) : unit  -- use default parameters
    this.file_name = fname
    this.mode = if mode.eq("") then "w+r+a" else mode end
    this.open()
  end

  def private open() : unit
    this.file = EMBED (FILE)
                  FILE *file = fopen(#{this.file_name.cstring}, #{this.mode.cstring});
                  file;
                END
    if not this.valid() then
      abort("Cannot open file, exiting.")
    end
  end

  def close() : unit
    if not this.valid() then
      abort("Cannot close file, exiting.\n")
    end
    this.file = EMBED (FILE)
                  fclose(#{this.file});
                  (FILE*)NULL;
                END
  end


  -- write a string
  def write(content:String) : unit
    if not this.valid() then
        abort("Cannot open file, exiting.")
    end
    EMBED (unit)
      fprintf(#{this.file}, #{content.cstring});
    END
  end

  def private valid() : bool
    EMBED (bool) (bool)#{this.file}; END
  end

  -- read a line
  def readline() : String
    var line = EMBED (CString)
                 char* line = encore_alloc(*_ctx,1024);
                 fgets(line, 1024, #{this.file});
                 line;
               END
    new String(line)
  end

  def eof() : bool
    EMBED (bool) feof(#{this.file}); END
  end
end


fun with_file[t](fname : FilePath, f : File -> t, mode : String = "r") : t
  val file  = new File(fname, mode)
  val res = f(file)
  file.close()
  res
end

-- this is a bad idea for large files because file is read in totality into memory
-- a lazy approach is required.
fun read_file(fname : FilePath) : String
  val file = new File(fname, "r")
  val sb = new StringBuffer()
  while not file.eof() do
    sb.add(file.readline())
  end
  file.close()
  sb.build()
end

fun write_file(fname : FilePath, str : String) : unit
  val file : File = new File(fname, "w")
  file.write(str)
  file.close()
end


fun append_file(fname : FilePath, str : String) : unit
  val file : File = new File(fname, "a")
  file.write(str)
  file.close()
end
