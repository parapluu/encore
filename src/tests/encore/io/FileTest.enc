import IO.File

active class Main
  val file : String = "FileTest.tmp"
  val datafile : String = "FileTestInput.txt"
  val msg : String = "Hello World!\n"

  def test_write() : unit
    val f = new File(this.file, "w")
    f.write(this.msg)
    f.close()  
  end

  def test_read() : unit
    val f = new File(this.file, "r")
    while (not f.eof()) do
      print("{}", f.readline())
    end
    f.close()
  end
 
  def test_with_file() : unit
    println("{}", with_file[int](this.datafile, lines))    
  end

  def test_write_file() : unit
    write_file(this.file, this.msg)
  end

  def test_append_file() : unit
    append_file(this.file, this.msg)  
  end

  def test_read_file() : unit
    print("{}", read_file(this.file))
  end

  def main () : unit
    -- the order matters!!
    this.test_write()
    this.test_read()
    this.test_with_file()
    this.test_write_file()
    this.test_append_file()
    this.test_read_file()
  end
end


fun lines(f : File) : int
  var i : int = 0
  while (not f.eof()) do
    f.readline()
    i += 1
  end
  i
end
