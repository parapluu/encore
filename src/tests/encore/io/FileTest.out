Hello World!
9
Hello World!
Hello World!
